### PR TITLE
Polish project documentation

### PR DESCRIPTION
**Pull request by osinstom**
This PR updates the project documentation, describing the current state of the project.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,91 +3,144 @@ SPDX-License-Identifier: Apache-2.0
...
+* UPF-initiated PFCP association  
+* UPF-based UE IP address assignment
+* Application filtering using the SDF filters
+* Sending of End Marker Packets
+* Downlink Data Notification (DDN) using PFCP Session Report
+* Integration with Prometheus for metrics about PFCP sessions or data plane level metrics. 
+* Application filtering using application PFDs (_**experimental**_).
 
-### Complete
+**BESS-UPF**
```
[the remark this answers is not in the document]
Comment (pudelkoM) [reply README.md:48]: What comes to mind are:

- Simple QoS support with slice and session meters
- Per-flow latency and throughout metrics

File: README.md [hunk 2]
```diff
@@ -3,91 +3,144 @@ SPDX-License-Identifier: Apache-2.0
...
+# upf
 
 [![Go Report Card](https://goreportcard.com/badge/github.com/omec-project/upf-epc)](https://goreportcard.com/report/github.com/omec-project/upf-epc)
 
 [![Build Status](https://jenkins.onosproject.org/buildStatus/icon?job=bess-upf-linerate-tests&subject=Linerate+Tests)](https://jenkins.onosproject.org/job/bess-upf-linerate-tests/)
 
-## Overview
+This project implements User Plane Function (UPF) - the 4G/5G mobile user plane compliant with 3GPP TS 23.501. 
+It follows the 3GPP CUPS (Control and User Plane Separation) architecture, making use of the PFCP protocol for the communication between SMF (5G) or SPGW-C (4G) and UPF.
+The UPF implementation is a part of [the Aether platform](https://opennetworking.org/aether/).
```
Comment (ccascone) [inline README.md:14]:
```suggestion
This project implements a 4G/5G User Plane Function (UPF) compliant with 3GPP TS 23.501. It follows the 3GPP Control and User Plane Separation (CUPS_ architecture, making use of the PFCP protocol for the communication between SMF (5G) or SPGW-C (4G) and UPF.
This UPF implementation is actively used as part of [the Aether platform](https://opennetworking.org/aether/) in conjunction with the SD-Core mobile core control plane. 
```

File: README.md [hunk 3]
```diff
@@ -3,91 +3,144 @@ SPDX-License-Identifier: Apache-2.0
...
+## Overview
 
-* Visualize your dataplane pipeline
+![UPF overview](./docs/images/upf-overview.jpg)
 
-![Pipeline](docs/images/pipeline.svg)
+The UPF implementation consists of two layers: 
 
-## Demo
+- **the PFCP Agent** (_pfcpiface_) implements the northbound interface of UPF and exposes the PFCP endpoint to the 4G/5G control plane.
```
Comment (ccascone) [inline README.md:22]:
```suggestion
- **PFCP Agent:** (_pfcpiface_) a Go-based implementation of the PFCP northbound API used by the mobile core control plane.
```

File: README.md [hunk 4]
```diff
@@ -3,91 +3,144 @@ SPDX-License-Identifier: Apache-2.0
...
+- **fastpath** implements a data plane of UPF. The PFCP Agent implements fastpath plugins that translate the 
+  PFCP semantics to the fastpath-specific data plane configuration. We currently support two fastpath implementations: 
+  - BESS-UPF - the UPF implementation that is build on top of [Berkeley Extensible Software Switch](https://github.com/NetSys/bess/) (BESS) programmable framework.
+    Please see the ONFConnect 2019 [talk](https://www.youtube.com/watch?v=fqJGWcwcOxE) for more details. You can also see demo videos [here](https://www.youtube.com/watch?v=KxK64jalKHw) and [here](https://youtu.be/rWnZuJeUWi4).
+  - [UP4](https://github.com/omec-project/up4) - the open-source P4-based UPF implementation, which is a part of the SD-Fabric project. Note that UP4 and P4-UPF names are exchangeable.
```
Comment (ccascone) [inline README.md:27]:
```suggestion
- **Fastpath:** responsible for the actual data plane packet processing.

The PFCP Agent implements fastpath plugins that translate 
  PFCP messages to fastpath-specific configurations. We currently support two fastpath implementations: 
  - BESS: a software-based fastpath built on top of the [Berkeley Extensible Software Switch](https://github.com/NetSys/bess/) (BESS) framework.
    Please see the ONFConnect 2019 [talk](https://www.youtube.com/watch?v=fqJGWcwcOxE) for more details. You can also see demo videos [here](https://www.youtube.com/watch?v=KxK64jalKHw) and [here](https://youtu.be/rWnZuJeUWi4).
  - [UP4](https://github.com/omec-project/up4) - an implementation leveraging ONOS and P4-programmable switches to realize a hardware-based fastpath. 
 
The combination of PFCP Agent and UP4 is usually referred to as P4-UPF. While BESS-UPF denotes the combination of PFCP Agent and the BESS fastpath.
```

File: README.md [hunk 5]
```diff
@@ -3,91 +3,144 @@ SPDX-License-Identifier: Apache-2.0
...
+  
+The UPF fastpaths are abstracted via the Fastpath API, which provides the means to communicate with the data plane.
+This design makes the UPF implementation extensible, enabling integration of new UPF fastpaths.
+Note that the PFCP Agent logic (e.g. handling of PFCP messages) is common and does need to be adjusted for new UPF fastpaths.
```
Comment (ccascone) [inline README.md:31]:
```suggestion
PFCP Agent internally abstracts different fastpaths using a common API, while the different plug-ins can use spcific southbound protocols to communicate with the different fastpath instances. Support for new fastpaths can be provided by implementing new plugins.
```

File: README.md [hunk 6]
```diff
@@ -3,91 +3,144 @@ SPDX-License-Identifier: Apache-2.0
...
+* The northbound PFCP interface including PFCP Association Setup/Release and Heartbeats 
+* Handling of the following PFCP entities: Packet Detection Rules (PDRs), Forwarding Action Rules (FARs),
+QoS Enforcement Rules (QERs).
+* UPF-initiated PFCP association  
+* UPF-based UE IP address assignment
+* Application filtering using the SDF filters
+* Sending of End Marker Packets
+* Downlink Data Notification (DDN) using PFCP Session Report
+* Integration with Prometheus for metrics about PFCP sessions or data plane level metrics. 
+* Application filtering using application PFDs (_**experimental**_).
```
Comment (ccascone) [inline README.md:46]:
```suggestion
* PFCP Association Setup/Release and Heartbeats 
* Session Establishment/Modification with support for PFCP entities such as Packet Detection Rules (PDRs), Forwarding Action Rules (FARs),
QoS Enforcement Rules (QERs).
* UPF-initiated PFCP association  
* UPF-based UE IP address assignment
* Application filtering using SDF filters
* Generation of End Marker Packets
* Downlink Data Notification (DDN) using PFCP Session Report
* Integration with Prometheus for exporting PFCP and data plane-level metrics. 
* Application filtering using application PFDs (_**experimental**_).
```

File: README.md [hunk 7]
```diff
@@ -3,91 +3,144 @@ SPDX-License-Identifier: Apache-2.0
...
+
+**P4-UPF**
+
+P4-UPF implements a core set of features capable of supporting requirements for a broad range of enterprise use cases.
+See [the ONF's blog post](https://opennetworking.org/news-and-events/blog/using-p4-and-programmable-switches-to-implement-a-4g-5g-upf-in-aether/) for an overview of P4-UPF. 
+Refer to [the SD-Fabric documentation](https://docs.sd-fabric.org/master/index.html) for the detailed feature set.
+
+## Getting started
+
+The UPF project provides two Docker images: `pfcpiface` (the PFCP Agent) and `bess` (the BESS-based UPF data plane).
```
Comment (ccascone) [inline README.md:75]:
```suggestion
This repository provides code to build two Docker images: `pfcpiface` (the PFCP Agent) and `bess` (the BESS-based fastpath). 
```

File: README.md [hunk 8]
```diff
@@ -3,91 +3,144 @@ SPDX-License-Identifier: Apache-2.0
...
+make docker-build
+```
+
+To build a selected image use `DOCKER_TARGETS`:
+
+```
+DOCKER_TARGETS=pfcpiface make docker-build
+```
+
+The latest Docker images are also published in [the OMEC project's Github registry](https://github.com/orgs/omec-project/packages?repo_name=upf).
```
Comment (ccascone) [inline README.md:89]: Let's use the Dopckerhub registry, which is what we use in helm charts

File: README.md [hunk 9]
```diff
@@ -3,91 +3,144 @@ SPDX-License-Identifier: Apache-2.0
...
-* SDF and APN based Qos Metering for MBR.
-* Sponsored Domain Name support
-* Buffering of downlink data
+To report any other kind of problem, feel free to open a GitHub Issue or reach out to the project maintainers on the ONF Community Slack.
 
-## Installation
+## License
 
-Please see [INSTALL.md](docs/INSTALL.md) for details on how to set up CP and UPF-EPC.
+The UPF implementation is licensed under the [Apache License, version 2.0](./LICENSES/Apache-2.0.txt).
```
Comment (ccascone) [inline README.md:146]:
```suggestion
This project is licensed under the [Apache License, version 2.0](./LICENSES/Apache-2.0.txt). 
```

**Review by ccascone — DISMISSED** [review]
Overall looks good. Please feel free to incorporate my suggestions as you see fit.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (osinstom) [reply README.md:49]: Added.

File: README.md [hunk 10]
```diff
@@ -3,91 +3,147 @@ SPDX-License-Identifier: Apache-2.0
...
+* Downlink Data Notification (DDN) - notification only (no buffering)
+* Basic QoS support, with per-slice and per-session rate limiting
+* Per-flow latency and throughput metrics  
+* Network Token Functions (_**experimental**_)
+
+**P4-UPF**
+
+P4-UPF implements a core set of features capable of supporting requirements for a broad range of enterprise use cases.
+See [the ONF's blog post](https://opennetworking.org/news-and-events/blog/using-p4-and-programmable-switches-to-implement-a-4g-5g-upf-in-aether/) for an overview of P4-UPF. 
+Refer to [the SD-Fabric documentation](https://docs.sd-fabric.org/master/index.html) for the detailed feature set.
```
Comment (ccascone) [inline README.md:74]:
```suggestion
Refer to [the SD-Fabric documentation](https://docs.sd-fabric.org/master/advanced/upf.html) for the detailed feature set.
```

File: README.md [hunk 11]
```diff
@@ -3,91 +3,147 @@ SPDX-License-Identifier: Apache-2.0
...
-![Pipeline](docs/images/pipeline.svg)
+- **PFCP Agent:** (_pfcpiface_) a Go-based implementation of the PFCP northbound API used to interact with the mobile core control plane.
+- **Fastpath:** responsible for the actual data plane packet processing.
 
-## Demo
+The PFCP Agent implements fastpath plugins that translate 
+  PFCP messages to fastpath-specific configurations. We currently support two fastpath implementations: 
+  - BESS: a software-based fastpath built on top of the [Berkeley Extensible Software Switch](https://github.com/NetSys/bess/) (BESS) framework.
+    Please see the ONFConnect 2019 [talk](https://www.youtube.com/watch?v=fqJGWcwcOxE) for more details. You can also see demo videos [here](https://www.youtube.com/watch?v=KxK64jalKHw) and [here](https://youtu.be/rWnZuJeUWi4).
+  - [UP4](https://github.com/omec-project/up4) - an implementation leveraging ONOS and P4-programmable switches to realize a hardware-based fastpath.
```
Comment (ccascone) [inline README.md:28]:
```suggestion
  - [UP4](https://github.com/omec-project/up4): an implementation leveraging ONOS and P4-programmable switches to realize a hardware-based fastpath. 
```

**Review by ccascone — APPROVED** [review]
More minor suggestions. It would be good to have an explanation of the content of each directory, but that go in a future PR. Feel free to merge when ready.

**Pull request merged**